### DISCUSSION TITLE
Advice on Deploying Harbor Registry in Active-Active Mode

### DISCUSSION THREAD

**Discussion by amitkumar2283** [goharbor/harbor#21255] [Q&A]
## TLDR:
Are there any recommended deployment patterns or best practices for deploying Harbor Registry in an active-active mode? 

## Long Read:
I'm reaching out to seek advice on deploying Harbor Registry in an active-active mode across multiple sites. Our goal is to ensure complete redundancy with each site having its own storage and database.

Our proposed architecture includes:

- Harbor Registry instances deployed in active-active mode across multiple sites
- Each site having its own storage (S3) and database
- Harbor instances behind a Load Balancer
- Bidirectional sync of storage (S3) to ensure images are synced across both Harbor instances

We're aware of potential challenges and would like to seek your advice on the following:

1. Unsynced DB data: Should we ensure that the databases are also in sync? If so, what's the recommended approach?
2. Robots and credentials: How can we ensure that robots and credentials created on one instance are synced with the other instance?
3. Caches: Are there any caching considerations we should be aware of in an active-active deployment?

We'd appreciate any guidance or advice you can offer.

Thank you for your time and expertise.

**Comment by stonezdj** [comment]
Answer to your questions:
1. Database should be synced.
2. Robot account and credential, they are stored in database. should be synced.
3. Cache can be unsynced.

What is the reason for having active-active? are you planing push image on both sites or just pull image from the second site?  could you please elaborate more detail on how you pull/push image on both sites?

**Reply by amitkumar2283** [reply]
Thanks for the response @stonezdj 

Use case:
We want to setup the registry for high uptime for both read and write, hence are looking for active-active for redundancy.

I was looking at Harbor Replication, which solves a few problems (around storage and project metadata syncing). But the task of synchronising tokens/credentials still remains (we will not be using LDAP/federated auth).

Following on your response, do we need to take anything else in consideration?

**Reply by omartrigui** [reply]
@amitkumar2283 Curious, did you manage to sync tokens and credentials? I intend to use Terraform to sync my local users and robot accounts by applying the changes to both active sides.
